src/fix/main.c

```diff
@@ -65,535 +75,946 @@ static void print_usage(void)
+	// Output padding
+	if (padValue != 0x100) {
+		if (input == output) {
+			if (lseek(output, 0, SEEK_END) == (off_t)-1) {
+				report("FATAL: Failed to seek to end of \"%s\": %s\n",
+				       name, strerror(errno));
+				goto free_romx;
+			}
+		}
+		memset(bank, padValue, sizeof(bank));
+		size_t len = (nbBanks - 1) * BANK_SIZE - totalRomxLen; // Don't count ROM0!
+
+		while (len) {
+			static_assert(sizeof(bank) <= SSIZE_MAX);
+			size_t thisLen = len > sizeof(bank) ? sizeof(bank) : len;
+			ssize_t ret = writeBytes(output, bank, thisLen);
+
+			if (ret != thisLen) {
+				report("FATAL: Failed to write \"%s\"'s padding: %s\n",
+				       name, strerror(errno));
+				break;
+			}
+			len -= thisLen;
+		}
 	}
 
-	if (setversion) {
-		/*
-		 * Offset 0x14C: Mask ROM Version Number
-		 * Which version of the ROM this is.
-		 */
+free_romx:
+	free(romx);
+}
+
+#undef trySeek
 
-		header[0x4C] = version;
+static bool processFilename(char const *name)
+{
+	nbErrors = 0;
+	if (name[0] == '-' && name[1] == '\0') {
+		name = "<stdin>";
+		processFile(STDIN_FILENO, STDOUT_FILENO, name, 0);
+
+	} else {
+		// POSIX specifies that the results of O_RDWR on a FIFO are undefined.
+		// However, this is necessary to avoid a TOCTTOU, if the file was changed between
+		// `stat()` and `open(O_RDWR)`, which could trigger the UB anyway.
+		// Thus, we're going to hope that either the `open` fails, or it succeeds but IO
+		// operations may fail, all of which we handle.
+		int input = open(name, O_RDWR);
+		struct stat stat;
+
+		if (input == -1) {
+			fprintf(stderr, "FATAL: Failed to open \"%s\" for reading+writing: %s\n",
+				name, strerror(errno));
+			nbErrors = 1;
+			goto fail;
+		}
+		if (fstat(input, &stat) == -1) {
+			close(input);
+			fprintf(stderr, "FATAL: Failed to stat \"%s\": %s\n",
+				name, strerror(errno));
+			nbErrors = 1;
+			goto fail;
+		}
+		// FIXME: Do we want to support other file types?
+		if (!S_ISREG(stat.st_mode)) {
+			close(input);
+			fprintf(stderr,
+				"FATAL: \"%s\" is not a regular file, and thus cannot be modified in-place\n",
+				name);
+			nbErrors = 1;
+			goto fail;
+		}
+
+		processFile(input, input, name, stat.st_size);
+
+		close(input);
 	}
+	if (nbErrors)
+fail:
+		fprintf(stderr, "Fixing \"%s\" failed with %u errors\n", name, nbErrors);
+	return nbErrors;
+}
 
-	if (fixheadsum || trashheadsum) {
-		/*
-		 * Offset 0x14D: Header Checksum
-		 */
+int main(int argc, char *argv[])
+{
+	char ch;
 
-		uint8_t headcksum = 0;
+	while ((ch = musl_getopt_long_only(argc, argv, optstring, longopts, NULL)) != -1) {
+		switch (ch) {
+			size_t len;
+#define parseByte(output, name) \
+do { \
+	char *endptr; \
+	unsigned long tmp = strtoul(optarg, &endptr, 0); \
+	\
+	if (!optarg[0] || *endptr) \
+		fprintf(stderr, "Expected number as argument to option '" name "', got %s\n", \
+			optarg); \
+	else if (tmp > 0xFF) \
+		fprintf(stderr, "Argument to option '" name "' is too large (%lu ≤ 255)\n", tmp); \
```

Comment:
ASCII `<=`, not UTF-8 (or whatever the .c file is encoded) `≤`.

Reply:
I'm not sure how to best convey that we *want* `tmp <= 255`. As for `<=`, why not? Are there popular terminal emulators that don't support UTF-8?

Reply:
There's this one OS that is called after a common household fixture. Doors? Ceilings? I don't quite recall the name, but it will come to me eventually.

Reply:
The user's terminal may not be using UTF-8 even if it is supported.

Reply:
Lowest common denominator it is, then; what about the conveyance issue?

Reply:
`Argument to option 'X' out of range 0-255: 257`
(Here it's unsigned so is definitely ≥ 0, but further below there's the `(0 ≤ %s ≤ 255)` message for the MBC ID which could match this one)